libraries/rush-lib/src/utilities/WebClient.ts

```diff
@@ -53,49 +56,124 @@ export enum WebClientProxy {
   Detect,
   Fiddler
 }
+export interface IRequestOptions extends RequestOptions, Pick<IFetchOptionsWithBody, 'body' | 'redirect'> {}
+
+export type FetchFn = (
+  url: string,
+  options: IRequestOptions,
+  isRedirect?: boolean
+) => Promise<IWebClientResponse>;
+
+const makeRequestAsync: FetchFn = async (
+  url: string,
+  options: IRequestOptions,
+  redirected: boolean = false
+) => {
+  const { body, redirect } = options;
+
+  return await new Promise(
+    (resolve: (result: IWebClientResponse) => void, reject: (error: Error) => void) => {
+      const parsedUrl: URL = typeof url === 'string' ? new URL(url) : url;
+      const requestFunction: typeof httpRequest | typeof httpsRequest =
+        parsedUrl.protocol === 'https:' ? httpsRequest : httpRequest;
+
+      requestFunction(url, options, (response: IncomingMessage) => {
+        const responseBuffers: (Buffer | Uint8Array)[] = [];
+        response.on('data', (chunk: string | Buffer | Uint8Array) => {
+          responseBuffers.push(Buffer.from(chunk));
+        });
+        response.on('end', () => {
+          // Handle retries by calling the method recursively with the redirect URL
+          const statusCode: number | undefined = response.statusCode;
+          if (statusCode === 301 || statusCode === 302) {
+            switch (redirect) {
+              case 'follow': {
+                const redirectUrl: string | string[] | undefined = response.headers.location;
+                if (redirectUrl) {
+                  makeRequestAsync(redirectUrl, options, true).then(resolve).catch(reject);
+                } else {
+                  reject(
+                    new Error(`Received status code ${response.statusCode} with no location header: ${url}`)
+                  );
+                }
+
+                break;
+              }
+              case 'error':
+                reject(new Error(`Received status code ${response.statusCode}: ${url}`));
+                return;
+            }
+          }
+
+          const responseData: Buffer = Buffer.concat(responseBuffers);
+          const status: number = response.statusCode || 0;
+          const statusText: string | undefined = response.statusMessage;
+          const result: IWebClientResponse = {
+            ok: status >= 200 && status < 300,
+            status,
+            statusText,
+            redirected,
+            getTextAsync: async () => responseData.toString(),
+            getJsonAsync: async () => JSON.parse(responseData.toString()),
```

Comment:
May make sense to memoize the result of `responseData.toString()` so it doesn't get called each time. Though ideally one would only need to call this method once anyway